pkg/gadget-service/service-oci.go

```diff
@@ -0,0 +1,208 @@
+
+	gi, err := s.runtime.GetOCIGadgetInfo(gadgetCtx, s.runtime.ParamDescs().ToParams(), req.ParamValues)
+	if err != nil {
+		return nil, fmt.Errorf("getting gadget info: %w", err)
+	}
+	return &api.GetOCIGadgetInfoResponse{GadgetInfo: gi}, nil
+}
+
+func (s *Service) RunOCIGadget(runGadget api.OCIGadgetManager_RunOCIGadgetServer) error {
+	ctrl, err := runGadget.Recv()
+	if err != nil {
+		return err
+	}
+
+	ociRequest := ctrl.GetOciRunRequest()
+	if ociRequest == nil {
+		return fmt.Errorf("expected first control message to be gadget run request")
+	}
+
+	if ociRequest.Version != api.VersionGadgetRunProtocol {
+		return fmt.Errorf("expected version to be %d, got %d", api.VersionGadgetRunProtocol, ociRequest.Version)
+	}
+
+	// Create a new logger that logs to gRPC and falls back to the standard logger when it failed to send the message
+	logger := logger.NewFromGenericLogger(&Logger{
+		send:           runGadget.Send,
+		level:          logger.Level(ociRequest.LogLevel),
+		fallbackLogger: s.logger,
+	})
+
+	runtime := s.runtime
+
+	for k, v := range ociRequest.ParamValues {
+		logger.Debugf("param %s: %s", k, v)
+	}
+
+	runtimeParams := s.runtime.ParamDescs().ToParams()
+	runtimeParams.CopyFromMap(ociRequest.ParamValues, "runtime.")
+
+	done := make(chan bool)
+	defer func() {
+		done <- true
+	}()
+
+	// Build a simple operator that subscribes to all events and forwards them
+	svc := simple.New("svc",
+		simple.WithPriority(50000),
+		simple.OnInit(func(gadgetCtx operators.GadgetContext) error {
+			// Create payload buffer
+			outputBuffer := make(chan *api.GadgetEvent, s.eventBufferLength)
+
+			log := gadgetCtx.Logger()
+
+			go func() {
+				// Receive control messages
+				for {
+					msg, err := runGadget.Recv()
+					if err != nil {
+						s.logger.Warnf("error on connection: %v", err)
+						gadgetCtx.Cancel()
+						return
+					}
+					switch msg.Event.(type) {
+					case *api.OCIGadgetControlRequest_StopRequest:
+						log.Debugf("received stop request")
+						gadgetCtx.Cancel()
+						return
+					default:
+						s.logger.Warn("received unexpected request")
+					}
+				}
+			}()
+
+			go func() {
+				// Message pump to handle slow readers
+				for {
+					select {
+					case ev := <-outputBuffer:
+						runGadget.Send(ev)
+					case <-done:
+						return
+					}
+				}
+			}()
+
+			seq := uint32(0)
+			var seqLock sync.Mutex
+
+			gi, err := gadgetCtx.SerializeGadgetInfo()
+			if err != nil {
+				return fmt.Errorf("serializing gadget info: %w", err)
+			}
+
+			// datasource mapping; we're sending an array of available DataSources including a
+			// DataSourceID; this ID will be used when sending actual data and needs to be remapped
+			// to the actual DataSource on the client later on
+			dsLookup := make(map[string]uint32)
+			for i, ds := range gi.DataSources {
+				ds.Id = uint32(i)
+				dsLookup[ds.Name] = ds.Id
+			}
+
+			// todo: skip DataSources we're not interested in
```

Comment:
I had that prepared, but I had to change it too often in the past weeks, so I postponed it. I still want to address it soon, but maybe - as it's an optimization - we should move it out of this PR.

This would fill out / read the IsRequested states. The CLIOperator has to set those (by subscribing to the DataSources on the client) and the service should decide that only the subscribed to DataSources should be sent. Same applies to the fields - much of the logic is present, but it's not yet 100%.